Add testset

Y4NG333 opened the pull request:
Add a test set to test the model with data that is different from the training set

Y4NG333 committed:
Added functionality for storing models 

1.Modified some training times and learning rate to make its loss decrease again. 
2.After the training is complete, the trained model is stored and used in the subsequent test set.

Y4NG333 committed:
add test set

Add a test set to test the model with data that is different from the training set

Y4NG333 commented:
@albert-no 교수님, 저는 main.py에 모형 저장 기능을 추가했습니다.
그리고 test.py 파일을 추가했습니다.
성능 테스트에서는 이미 훈련된 model을 도입하여 테스트할 수 있다.

albert-no commented on test.py:
```diff
@@ -0,0 +1,57 @@
+os.environ["KMP_DUPLICATE_LIB_OK"] = "TRUE"
+
+alphabet = ["a", "b", "c"]
+weight = np.array([1, 1, 1])
```
main에서는 2, 1, 1을 썼습니다.

Y4NG333 replied:
숫자를 입력할 때 실수로 오타를 냈습니다.
지금 제가 수정했습니다.

albert-no commented on test.py:
```diff
@@ -0,0 +1,57 @@
+# Load model parameters
+path_model = "./model/"
+model_test = Transformer(n_heads, d_model, n_layers, src_vocab_size, tgt_vocab_size)
+model_test.load_state_dict(torch.load(path_model + str(20) + "net.pth"))
```
path_model+str(20) .. 과 같이 쓰지말고
애초에 model.pth이름을 `fname`으로 입력을 받기 바랍니다.

Y4NG333 replied:
교수님, 저는 잘 모르겠습니다... fname으로 이 무엇입니까?

albert-no replied:
```
fname = "./model/20net.pth"
model_test.load_state_dict(torch.load(fname))
```
과 같이 pth파일이름을 한꺼번에 저장하는 변수를 활용하기 바랍니다

Y4NG333 replied:
네, 알겠습니다!

albert-no commented on test.py:
```diff
@@ -0,0 +1,57 @@
+model_test.load_state_dict(torch.load(path_model + str(20) + "net.pth"))
+
+# Generate test set
+seq, seq_int, code, code_int, code_onehot = data_generator(datainfo, 10)
```
여기 10은 뭔가요? 변수를 만들기 바랍니다.
이렇게 직접적으로 숫자를 사용하면 나중에 코드를 수정하기 어려워집니다.

Y4NG333 replied:
10은 train set의 수를 나타냅니다.
저는 10을 대체할 변수를 만들었습니다.

albert-no commented on test.py:
```diff
@@ -0,0 +1,57 @@
+outputs, enc_self_attns, dec_self_attns, dec_enc_attns = model_test(seq_int, code_int)
+real_out = [torch.argmax(x).item() for x in outputs]
+
+print("real out", np.array(real_out).reshape(-1, 12))
```
12도 max_len에 해당하는건가요?

Y4NG333 replied:
네, 맞습니다.12는 최대 길이(max_len)입니다.

albert-no replied:
여기도 12가 아니라 max_len을 활용하기 바랍니다.
```
print("real out", np.array(real_out).reshape(-1, max_len))
```

Y4NG333 replied:
네, 교수님, 수정했습니다.

albert-no commented on test.py:
```diff
@@ -0,0 +1,57 @@
+
+print("real out", np.array(real_out).reshape(-1, 12))
+print("code int", code_int)
+loss = criterion(outputs.float(), torch.LongTensor(code_onehot).view(-1, 3).float())
```
여기서 3은 뭐죠?

Y4NG333 replied:
출력은 1*3의 vector이기 때문에 loss를 계산할 때 출력과 같은 차원을 유지합니다.
![0BF3A80F45A5FB0E58A7E65151A72802](https://user-images.githubusercontent.com/91429283/160230828-41abb34d-eb03-4abf-ac33-06716e1d38b6.jpg)

albert-no commented on test.py:
```diff
@@ -0,0 +1,57 @@
+weight = np.array([1, 1, 1])
+prob = weight / np.sum(weight)
+
+num_epoch = 21
```
epoch이나 batch는 test에서 필요하지 않습니다.
불필요한 변수들은 지우기 바랍니다.

Y4NG333 replied:
코드의 일부는 main.py에서 직접 복사하기 때문에 관련이 없는 변수들이 있습니다...
저는 이제 관련 없는 변수들을 모두 삭제했습니다.

albert-no reviewed (commented):
코멘트 확인하여 수정하기 바랍니다.

Y4NG333 commented:
교수님，안녕하세요~
네, 빨리 수정하겠습니다.

Y4NG333 committed:
np.array([1, 1, 1]) -> np.array([2, 1, 1])

Y4NG333 committed:
필요 없는 변수 삭제

Y4NG333 committed:
Replace the number 10 with a variable

Y4NG333 commented:
@albert-no  교수님, 정말 죄송합니다. 내 컴퓨터 결함 때문에 (컴퓨터에서 자꾸 이상한 소리가 난다)이 변경 사항들은 오늘 오후에야 제출합니다.
아직 제가 잘하지 못하는 곳이 하나 있습니다.저는 model.pth이름을 무엇으로 수정해야 할지 잘 모르겠습니다.

albert-no commented:
코멘트 확인하기 바랍니다.
그리고 숫자 (3이나 12)를 직접활용하기 보다 변수를 활용하기 바랍니다.
그리고 commit을 한글로 하지말고 영어로 하기 바랍니다.

Y4NG333 committed:
using max_len variables

Y4NG333 committed:
Use fname for model.pth

Y4NG333 committed:
using dimension variables

3 --> dimension

Y4NG333 commented:
@albert-no 교수님, 관련 부분은 전부 수정했습니다.
아마 그때 입력 방법이 한글이었나봐요, 그래서 한글로 commit을 입력했어요. 
다음에는 전부 영어로 하겠습니다.

Y4NG333 committed:
Use the learning rate scheduler

After using the learning rate scheduler, the loss drops from about 0.28 to below 0.07. 
Modified some variable names and variable values .

Y4NG333 committed:
Modify model name
